### PR TITLE
Implement case insensitive (and negated) regex.

### PR DESCRIPTION
**Pull request by sellout**
Also bump minimum scoverage req to 80%.

### CODE REVIEW

**Comment by jonsterling** [comment]
Heh, thanks for getting this in!

**Comment by mossprescott** [comment]
We're adding three new, unguessable operators (`~*`, `!~`, and `!~*`), which are at least common with Postgres, and we're adding a new argument to our own `SEARCH` function. Where is this query language documented? For one thing, the cheatsheet should get updated (see https://github.com/slamdata/slamdata.github.io/blob/master/support/cheatsheet.pdf and https://slamdata.atlassian.net/browse/SD-1075).

One possible answer would be to punt that to SQL^2 (and maybe write an issue for creating that documentation).

File: core/src/main/scala/quasar/physical/mongodb/planner.scala [hunk 1]
```diff
@@ -203,9 +210,23 @@ object MongoDbPlanner extends Planner[Crystallized] with Conversions {
           Arity3((field, start, len) =>
             Call(Select(field, "substr"), List(start, len)))
         case Search =>
-          Arity2((field, pattern) =>
-            Call(Select(New(Name("RegExp"), List(pattern)), "test"),
-              List(field)))
+          args match {
+            case List(a1, a2, a3) => (HasJs(a1) |@| HasJs(a2) |@| HasBool(a3)) {
```
Comment (mossprescott) [inline core/src/main/scala/quasar/physical/mongodb/planner.scala]: Losing me here. Can't this just be `List(HasJs(a1), ...)` or `HasJs(a1) :: ... :: Nil`?

Comment (mossprescott) [inline core/src/main/scala/quasar/physical/mongodb/planner.scala]: More importantly, since we don't seem to require the pattern to be a literal, why require the case-insensitivity flag to be a literal? It seems like we should avoid that kind of restriction in general if we can.

Comment (sellout) [inline core/src/main/scala/quasar/physical/mongodb/planner.scala]: Nope, this isn’t `workflowƒ`. And `Has*` aren’t extractors in `jsExprƒ`.

File: core/src/main/scala/quasar/physical/mongodb/planner.scala [hunk 2]
```diff
@@ -496,9 +526,7 @@ object MongoDbPlanner extends Planner[Crystallized] with Conversions {
         case (And, _)      => invoke2Nel(Selector.And.apply _)
         case (Or, _)       => invoke2Nel(Selector.Or.apply _)
         case (Not, (_, v) :: Nil) =>
-          v.map { case (sel, loc) =>
-            (sel andThen (s => s.negate)) -> loc.map(there(0, _))
-          }
+          v.map(_.bimap(_ andThen (_.negate), _.map(there(0, _))))
```
Comment (mossprescott) [inline core/src/main/scala/quasar/physical/mongodb/planner.scala]: Thanks for cleaning up my mess :)

File: core/src/main/scala/quasar/physical/mongodb/planner.scala [hunk 3]
```diff
@@ -545,13 +573,16 @@ object MongoDbPlanner extends Planner[Crystallized] with Conversions {
...
         selCheck(typ).fold[OutputM[PartialSelector[B]]](
           -\/(UnsupportedPlan(node, None)))(
-          f =>
-          \/-(cont._2.fold[PartialSelector[B]](
-            κ(({ case List(field) => f(field) }, List(there(0, here)))),
-            { case (f2, p2) =>
-              ({ case head :: tail => Selector.And(f(head), f2(tail)) },
-                there[B](0, here) :: p2.map(there(1, _)))
-            })))
+          f => cont._2.flatMap { case (f2, p2) =>
```
Comment (mossprescott) [inline core/src/main/scala/quasar/physical/mongodb/planner.scala]: How does this relate to the rest of the changes?

File: core/src/main/scala/quasar/physical/mongodb/planner.scala (hunk 1 above)
Comment (sellout) [inline core/src/main/scala/quasar/physical/mongodb/planner.scala]: That’s a good point. I was was mimicking what I did for Selectors, but clearly we have more flexibility in JS.

File: core/src/main/scala/quasar/std/string.scala [hunk 4]
```diff
@@ -56,18 +56,19 @@ trait StringLib extends Library {
     },
     basicUntyper)
 
-  def matchAnywhere(str: String, pattern: String) = java.util.regex.Pattern.compile(pattern).matcher(str).find()
+  def matchAnywhere(str: String, pattern: String, insen: Boolean) =
+    java.util.regex.Pattern.compile(if (insen) "(?i)" ⊹ pattern else pattern).matcher(str).find()
 
   val Search = Mapping(
     "search",
     "Determines if a string value matches a regular expresssion.",
```
Comment (mossprescott) [inline core/src/main/scala/quasar/std/string.scala]: This is probably the place to include some text about the parameters.

**Comment by mossprescott** [comment]
I think we're missing some integration tests:
* One or two for `!~` and `!~*` for completeness (the point being that every feature is represented by a test at that level, for documentation and future backend verification).
* One for `~*` used in non-selector position, as in `select city ~* 'bou.*, city from zips'. Can I assume we have a corresponding test for `~`?

File: core/src/main/scala/quasar/physical/mongodb/planner.scala (hunk 3 above)
Comment (sellout) [inline core/src/main/scala/quasar/physical/mongodb/planner.scala]: This was an accidental commit – should be in a separate PR.

File: it/src/main/resources/tests/guardedExpression.test [hunk 5]
```diff
@@ -0,0 +1,7 @@
+{
+  "name": "regex on non-string field",
+  "data": "zips.data",
```
Comment (mossprescott) [inline it/src/main/resources/tests/guardedExpression.test]: Here's where you broke the build. :-)

**Comment by sellout** [comment]
@mossprescott Updated (since you get no other notification).

**Comment by mossprescott** [comment]
Looking good now, and I like the cleanup of `jsExpr`. :+1:
